src/Dialect/ONNX/ONNXOps/NN/Attention.cpp

```diff
@@ -0,0 +1,210 @@
+    return onnx_mlir::Diagnostic::emitOperandHasUnexpectedRankError(
+        *this->getOperation(), q, qRank, "3 or 4");
+
+  // check q_num_heads is present for 3D input
+  auto qNumHeads = adaptor.getQNumHeads();
+  if (qRank == 3 && !qNumHeads)
+    return emitOpError("attribute 'q_num_heads' must be provided when input "
+                       "'q' is a 3D tensor.");
+
+  Value k = adaptor.getK();
+  Value v = adaptor.getV();
+  if (!hasShapeAndRank(k) || !hasShapeAndRank(v))
+    return success(); // Won't be able to do any more checking at this stage.
+
+  auto kType = mlir::cast<ShapedType>(k.getType());
+  int64_t kRank = kType.getShape().size();
+  if (kRank != 3 && kRank != 4)
+    return onnx_mlir::Diagnostic::emitOperandHasUnexpectedRankError(
+        *this->getOperation(), k, kRank, "3 or 4");
+
+  auto vType = mlir::cast<ShapedType>(v.getType());
+  int64_t vRank = vType.getShape().size();
+  if (vRank != 3 && vRank != 4)
+    return onnx_mlir::Diagnostic::emitOperandHasUnexpectedRankError(
+        *this->getOperation(), v, vRank, "3 or 4");
+
+  // check kv_num_heads is present for 3D inputs
+  auto kvNumHeads = adaptor.getKvNumHeads();
+  if ((kRank == 3 || vRank == 3) && !kvNumHeads)
+    return emitOpError("attribute 'kv_num_heads' must be provided when inputs "
+                       "'k' or 'v' are 3D tensors.");
+
+  auto divisibleByNumHeads =
+      [&](ShapedType &type, std::optional<int64_t> numHeads, Value &operand) {
+        if (type.hasStaticShape()) {
+          auto shape = type.getShape();
+          if (type.getRank() == 3 && numHeads && shape[2] % *numHeads != 0)
+            return onnx_mlir::Diagnostic::emitDimensionHasUnexpectedValueError(
+                *this->getOperation(), operand, 2, shape[2],
+                "divisible by " + std::to_string(*numHeads));
+        }
+        return success();
+      };
+
+  auto qTypeDivisibleByQNumHeads = divisibleByNumHeads(qType, qNumHeads, q);
+  if (!succeeded(qTypeDivisibleByQNumHeads))
+    return qTypeDivisibleByQNumHeads;
+
+  auto kTypeDivisibleByKVNumHeads = divisibleByNumHeads(kType, kvNumHeads, k);
+  if (!succeeded(kTypeDivisibleByKVNumHeads))
+    return kTypeDivisibleByKVNumHeads;
+
+  auto vTypeDivisibleByKVNumHeads = divisibleByNumHeads(vType, kvNumHeads, v);
+  if (!succeeded(vTypeDivisibleByKVNumHeads))
+    return vTypeDivisibleByKVNumHeads;
+
+  return success();
+}
+
+//===----------------------------------------------------------------------===//
+// Shape Inference
+//===----------------------------------------------------------------------===//
+
+LogicalResult ONNXAttentionOp::inferShapes(
+    std::function<void(Region &)> /*shapeInferenceFunc*/) {
+
+  // Cannot infer the output shape if the q, k, v operands' shapes are not yet
+  // known.
+  for (size_t i = 0; i < 3; i++)
+    if (!hasShapeAndRank(this->getOperand(i)))
+      return success();
+
+  Type elementType = mlir::cast<ShapedType>(getQ().getType()).getElementType();
```

Comment:
nit, no change needed:
There is
```suggestion
  Type elementType =getElementTypeOrSelf(getQ().getType());
```

this has the advantage that it also works with scalars, but for Attention this should not make any difference